File: internal/api/jdwp/method.go
package jdwp

import (
	"context"
	"fmt"

	"cli-debugger/internal/api"
)

// Method Command Set Implementation
// Method command constants are already defined in protocol.go

// LineLocation Line location information
type LineLocation struct {
	LineCodeIndex int64
	LineNumber    int
}

// VariableInfo Variable information
type VariableInfo struct {
	Slot      int
	Name      string
	Signature string
	CodeIndex int64
}

// LineTable Get method line table
func (c *Client) LineTable(refTypeID string, methodID string) ([]*LineLocation, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)
	data = append(data, encodeID(methodID, c.idsizes.MethodIDSize)...)

	packet := createCommandPacketWithData(methodCommandSet, methodCommandLineTable, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method line table",
			Cause:   err,
		}
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method line table",
			Cause:   err,
		}
	}

	if reply.ErrorCode != 0 {
		return nil, &api.APIError{
			Type:    api.ProtocolError,
			Code:    int(reply.ErrorCode),
			Message: fmt.Sprintf("Get method line table failed: %s", reply.Message),
		}
	}

	reader := newPacketReader(reply.Data)
	start := reader.readInt64()
	end := reader.readInt64()
	lines := reader.readInt()

	lineLocations := make([]*LineLocation, 0, lines)
	for i := 0; i < lines; i++ {
		lineCodeIndex := reader.readInt64()
		lineNumber := reader.readInt()

		lineLocations = append(lineLocations, &LineLocation{
			LineCodeIndex: lineCodeIndex,
			LineNumber:    lineNumber,
		})
	}

	_ = start
	_ = end

	return lineLocations, nil
}

// VariableTable Get method variable table
func (c *Client) VariableTable(refTypeID string, methodID string) ([]*VariableInfo, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)
	data = append(data, encodeID(methodID, c.idsizes.MethodIDSize)...)

	packet := createCommandPacketWithData(methodCommandSet, methodCommandVariableTable, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method variable table",
			Cause:   err,
		}
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method variable table",
			Cause:   err,
		}
	}

	if reply.ErrorCode != 0 {
		return nil, &api.APIError{
			Type:    api.ProtocolError,
			Code:    int(reply.ErrorCode),
			Message: fmt.Sprintf("Get method variable table failed: %s", reply.Message),
		}
	}

	reader := newPacketReader(reply.Data)
	argCount := reader.readInt()
	slots := reader.readInt()

	variableInfos := make([]*VariableInfo, 0, slots)
	for i := 0; i < slots; i++ {
		codeIndex := reader.readInt64()
		name := reader.readString()
		signature := reader.readString()
		slot := reader.readInt()
		length := reader.readInt()

		variableInfos = append(variableInfos, &VariableInfo{
			Slot:      slot,
			Name:      name,
			Signature: signature,
			CodeIndex: codeIndex,
		})

		_ = length
	}

	_ = argCount

	return variableInfos, nil
}

// Bytecodes Get method bytecodes
func (c *Client) Bytecodes(refTypeID string, methodID string) ([]byte, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)
	data = append(data, encodeID(methodID, c.idsizes.MethodIDSize)...)

	packet := createCommandPacketWithData(methodCommandSet, methodCommandBytecodes, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method bytecodes",
			Cause:   err,
		}
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method bytecodes",
			Cause:   err,
		}
	}

	if reply.ErrorCode != 0 {
		return nil, &api.APIError{
			Type:    api.ProtocolError,
			Code:    int(reply.ErrorCode),
			Message: fmt.Sprintf("Get method bytecodes failed: %s", reply.Message),
		}
	}

	reader := newPacketReader(reply.Data)
	bytecodes := reader.readBytes()

	return bytecodes, nil
}

// IsObsolete Check if method is obsolete
func (c *Client) IsObsolete(refTypeID string, methodID string) (bool, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)
	data = append(data, encodeID(methodID, c.idsizes.MethodIDSize)...)

	packet := createCommandPacketWithData(methodCommandSet, methodCommandIsObsolete, data)
	if err := c.sendPacket(packet); err != nil {
		return false, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to check if method is obsolete",
			Cause:   err,
		}
	}

	reply, err := c.readReply()
	if err != nil {
		return false, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to check if method is obsolete",
			Cause:   err,
		}
	}

	if reply.ErrorCode != 0 {
		return false, &api.APIError{
			Type:    api.ProtocolError,
			Code:    int(reply.ErrorCode),
			Message: fmt.Sprintf("Check if method is obsolete failed: %s", reply.Message),
		}
	}

	reader := newPacketReader(reply.Data)
	isObsolete := reader.readByte() != 0

	return isObsolete, nil
}

// VariableTableWithGeneric Get method variable table with generic signature
func (c *Client) VariableTableWithGeneric(refTypeID string, methodID string) ([]*VariableInfo, error) {
	data := make([]byte, 0)

	data = append(data, encodeID(refTypeID, c.idsizes.ReferenceTypeIDSize)...)
	data = append(data, encodeID(methodID, c.idsizes.MethodIDSize)...)

	packet := createCommandPacketWithData(methodCommandSet, methodCommandVariableTableWithGeneric, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method variable table with generic",
			Cause:   err,
		}
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get method variable table with generic",
			Cause:   err,
		}
	}

	if reply.ErrorCode != 0 {
		return nil, &api.APIError{
			Type:    api.ProtocolError,
			Code:    int(reply.ErrorCode),
			Message: fmt.Sprintf("Get method variable table with generic failed: %s", reply.Message),
		}
	}

	reader := newPacketReader(reply.Data)
	argCount := reader.readInt()
	slots := reader.readInt()

	variableInfos := make([]*VariableInfo, 0, slots)
	for i := 0; i < slots; i++ {
		codeIndex := reader.readInt64()
		name := reader.readString()
		signature := reader.readString()
		genericSignature := reader.readString()
		slot := reader.readInt()
		length := reader.readInt()

		variableInfos = append(variableInfos, &VariableInfo{
			Slot:      slot,
			Name:      name,
			Signature: signature,
			CodeIndex: codeIndex,
		})

		_ = genericSignature
		_ = length
	}

	_ = argCount

	return variableInfos, nil
}

File: src/protocol/jdwp/method.ts
/**
 * Method Command Set Implementation
 */

import {
  CommandSet,
  MethodCommand,
  type IDSizes,
  type LineLocation,
  type VariableInfo,
} from "./protocol.js";
import { createCommandPacketWithData, encodeID } from "./codec.js";
import { PacketReader } from "./reader.js";
import { APIError, ErrorType, ErrorCodes } from "../errors.js";

/**
 * JDWP Client interface for command execution
 */
export interface JDWPCommandExecutor {
  sendPacket(packet: Buffer): Promise<void>;
  readReply(): Promise<{ errorCode: number; message: string; data: Buffer }>;
  idSizes: IDSizes;
}

/**
 * Get method line table
 */
export async function getLineTable(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  methodID: string,
): Promise<LineLocation[]> {
  const data = Buffer.concat([
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeID(methodID, executor.idSizes.methodIDSize),
  ]);

  const packet = createCommandPacketWithData(
    CommandSet.Method,
    MethodCommand.LineTable,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get line table failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  reader.readInt64(); // start
  reader.readInt64(); // end
  const lines = reader.readInt();

  const lineLocations: LineLocation[] = [];
  for (let i = 0; i < lines; i++) {
    lineLocations.push({
      lineCodeIndex: reader.readInt64(),
      lineNumber: reader.readInt(),
    });
  }

  return lineLocations;
}

/**
 * Get method variable table
 */
export async function getVariableTable(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  methodID: string,
): Promise<VariableInfo[]> {
  const data = Buffer.concat([
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeID(methodID, executor.idSizes.methodIDSize),
  ]);

  const packet = createCommandPacketWithData(
    CommandSet.Method,
    MethodCommand.VariableTable,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get variable table failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  reader.readInt(); // argCount
  const slots = reader.readInt();

  const variableInfos: VariableInfo[] = [];
  for (let i = 0; i < slots; i++) {
    const codeIndex = reader.readInt64();
    const name = reader.readString();
    const signature = reader.readString();
    const slot = reader.readInt();
    reader.readInt(); // length

    variableInfos.push({
      slot,
      name,
      signature,
      codeIndex,
    });
  }

  return variableInfos;
}

/**
 * Get method bytecodes
 */
export async function getBytecodes(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  methodID: string,
): Promise<Buffer> {
  const data = Buffer.concat([
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeID(methodID, executor.idSizes.methodIDSize),
  ]);

  const packet = createCommandPacketWithData(
    CommandSet.Method,
    MethodCommand.Bytecodes,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get bytecodes failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readBytes();
}

/**
 * Check if method is obsolete
 */
export async function isObsolete(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  methodID: string,
): Promise<boolean> {
  const data = Buffer.concat([
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeID(methodID, executor.idSizes.methodIDSize),
  ]);

  const packet = createCommandPacketWithData(
    CommandSet.Method,
    MethodCommand.IsObsolete,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Check obsolete failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readByte() !== 0;
}

/**
 * Get method variable table with generic signature
 */
export async function getVariableTableWithGeneric(
  executor: JDWPCommandExecutor,
  refTypeID: string,
  methodID: string,
): Promise<VariableInfo[]> {
  const data = Buffer.concat([
    encodeID(refTypeID, executor.idSizes.referenceTypeIDSize),
    encodeID(methodID, executor.idSizes.methodIDSize),
  ]);

  const packet = createCommandPacketWithData(
    CommandSet.Method,
    MethodCommand.VariableTableWithGeneric,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get variable table with generic failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  reader.readInt(); // argCount
  const slots = reader.readInt();

  const variableInfos: VariableInfo[] = [];
  for (let i = 0; i < slots; i++) {
    const codeIndex = reader.readInt64();
    const name = reader.readString();
    const signature = reader.readString();
    reader.readString(); // genericSignature
    const slot = reader.readInt();
    reader.readInt(); // length

    variableInfos.push({
      slot,
      name,
      signature,
      codeIndex,
    });
  }

  return variableInfos;
}
